hs-bindgen/src/HsBindgen/C/Tc/Macro.hs

```diff
@@ -245,172 +300,238 @@ data MetaTyVar
   deriving anyclass PrettyVal
 
 type TyCon :: Nat -> Kind -> Hs.Type
-data TyCon n k where
-  DataTyCon  :: Nat.SNatI n => !( DataTyCon  n ) -> TyCon n Ty
-  ClassTyCon :: Nat.SNatI n => !( ClassTyCon n ) -> TyCon n Ct
+data TyCon nbArgs res where
+  GenerativeTyCon :: !( GenerativeTyCon nbArgs ki ) -> TyCon nbArgs ki
+  FamilyTyCon     :: !( FamilyTyCon     nbArgs    ) -> TyCon nbArgs Ty
+deriving stock instance Eq ( TyCon nbArgs res )
+
+type GenerativeTyCon :: Nat -> Kind -> Hs.Type
+data GenerativeTyCon nbArgs res where
+  DataTyCon      :: !( DataTyCon   nbArgs ) -> GenerativeTyCon nbArgs Ty
+  ClassTyCon     :: !( ClassTyCon  nbArgs ) -> GenerativeTyCon nbArgs Ct
+deriving stock instance Eq ( GenerativeTyCon nbArgs res )
+
+instance PrettyVal ( TyCon nbArgs k ) where
+  prettyVal ( GenerativeTyCon tc  ) = Pretty.Con "GenerativeTyCon" [ prettyVal tc  ]
+  prettyVal ( FamilyTyCon     fam ) = Pretty.Con     "FamilyTyCon" [ prettyVal fam ]
 
-instance PrettyVal ( TyCon n k ) where
-  prettyVal ( DataTyCon dc ) = Pretty.Con "DataTyCon" [ prettyVal dc ]
-  prettyVal ( ClassTyCon cls ) = Pretty.Con "ClassTyCon" [ prettyVal cls ]
+instance PrettyVal ( GenerativeTyCon nbArgs k ) where
+  prettyVal ( DataTyCon   dc  ) = Pretty.Con   "DataTyCon" [ prettyVal dc  ]
+  prettyVal ( ClassTyCon  cls ) = Pretty.Con  "ClassTyCon" [ prettyVal cls ]
 
 type DataTyCon :: Nat -> Hs.Type
-data DataTyCon n where
-  -- | Type constructor for Bool
+data DataTyCon nbArgs where
+  -- | Type constructor for 'Void'
+  VoidTyCon      :: DataTyCon Z
+  -- | Type constructor for 'Bool'
   BoolTyCon      :: DataTyCon Z
-  -- | Type constructor forString
+  -- | Type constructor for 'String'
   StringTyCon    :: DataTyCon Z
-  -- | Type constructors for integral types, such as 'Int' or 'UShort'.
-  IntLikeTyCon   :: !PrimIntType -> DataTyCon Z
+  -- | Type constructor for integral types, such as 'Int' or 'UShort'.
+  IntLikeTyCon   :: DataTyCon ( S Z )
   -- | Type constructor for floating-point types, such as 'Float' or 'Double'.
-  FloatLikeTyCon :: !PrimFloatType -> DataTyCon Z
+  FloatLikeTyCon :: DataTyCon ( S Z )
+  -- | Type constructor for pointers
+  PtrTyCon       :: DataTyCon ( S Z )
+
+  -- | Type constructor for the type of primitive integral types,
+  -- which appear as arguments to 'IntLikeTyCon'.
+  PrimIntTyCon   :: ( PrimIntType, PrimSign ) -> DataTyCon Z
```

Comment:
Use a different name for `PrimIntTyCon` e.g. `PrimIntShape` or something.

Also update the comment: "Type constructor for the type-level pair of a width and sign, intended for use with `IntLike`"